op-program/host/l2_source.go

```diff
@@ -0,0 +1,139 @@
+package host
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum-optimism/optimism/op-program/host/config"
+	"github.com/ethereum-optimism/optimism/op-program/host/prefetcher"
+	"github.com/ethereum-optimism/optimism/op-service/client"
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+	"github.com/ethereum-optimism/optimism/op-service/sources"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+// L2Source is a source of L2 data, it abstracts away the details of how to fetch L2 data between canonical and experimental sources.
+// It also tracks metrics for each of the apis. Once experimental sources are stable, this will only route to the "experimental" source.
+type L2Source struct {
+	logger log.Logger
+
+	// canonical source, used as a fallback if experimental source is enabled but fails
+	// the underlying node should be a geth hash scheme archival node.
+	canonicalEthClient   *L2Client
+	canonicalDebugClient *sources.DebugClient
+
+	// experimental source, used as the primary source if enabled
+	experimentalClient *L2Client
+
+	// whether to use the experimental source
+	experimentalEnabled bool
+}
+
+var _ prefetcher.L2Source = &L2Source{}
+
+// NewL2SourceWithClient creates a new L2 source with the given client as the canonical client.
+// This doesn't configure the experimental source, but is useful for testing.
+func NewL2SourceWithClient(logger log.Logger, canonicalL2Client *L2Client, canonicalDebugClient *sources.DebugClient) *L2Source {
+	source := &L2Source{
+		logger:               logger,
+		canonicalEthClient:   canonicalL2Client,
+		canonicalDebugClient: canonicalDebugClient,
+	}
+
+	return source
+}
+
+func NewL2Source(ctx context.Context, logger log.Logger, config *config.Config) (*L2Source, error) {
+	logger.Info("Connecting to canonical L2 source", "url", config.L2URL)
+
+	// eth_getProof calls are expensive and takes time, so we use a longer timeout
```

Comment:
Yep this is what we found (with a 2 second block time):

- block 12 hrs ago: execution witness took 41s, account proof took 1m9s
- block 24 hrs ago: execution witness took 1m48s, account proof took 4m
- block 6 days ago: execution witness took 3m33s, account proof took 8m

We could make the overall timeout configurable (maybe just across all RPC calls for simplicity)? Wdyt?

Reply:
I'd be looking at what it takes for a block from at least 14 and maybe even 30 days ago to be honest.  6 days is definitely too short given the air gap will mean the anchor state becomes a minimum of 7 days old. You can imagine if there's an issue with proposals that means we need to to intervene, worst case is that we intervene at the 7 day mark, but then we fix the issue and it will take another 7 days for a correct proposal to finalise and update the anchor state, so you wind up with it being 14 days old. 

Probably is worth making it configurable, though we'll then need to add an option to op-challenger so that it can pass through a config option to op-program. I suspect in practice something will need to be done to improve reth's performance here though or it opens up an avenue for an attacker to overwhelm the honest actor's resources by making them compute expensive proofs from old blocks.

Reply:
Created https://github.com/ethereum-optimism/optimism/issues/12678 to track making this configurable.